### ISSUE TITLE
Not able to install Picker with (latest) SDK 46 

### ISSUE THREAD

**Issue by ecthiender** [expo/expo#19068]
### Summary

I initialized a new Expo project with `create-expo-app` (SDK 46). I want to use the Picker component. I am trying to install it via `expo install @react-native-picker/picker` as mentioned in https://docs.expo.dev/versions/latest/sdk/picker/

### Expected Behaviour
Install successfully, and I'm able to `Picker` in my project.

### Actual Behaviour
I get an error as `@react-native-picker/picker` requires `react@"16 || 17"`

```
❯ npx expo install @react-native-picker/picker                                                                                                         
› Installing 1 SDK 46.0.0 compatible native module using npm                                                                                           
> npm install                                                                                                                                          
npm ERR! code ERESOLVE                                                                                                                                 
npm ERR! ERESOLVE unable to resolve dependency tree                                                                                                    
npm ERR!                                                                                                                                               
npm ERR! While resolving: my-app@1.0.0                                                                                                                 
npm ERR! Found: react@18.0.0                                                                                                                           
npm ERR! node_modules/react                                                                                                                            
npm ERR!   react@"18.0.0" from the root project                                                                                                        
npm ERR!                                                                                                                                               
npm ERR! Could not resolve dependency:                                                                                                                 
npm ERR! peer react@"16 || 17" from @react-native-picker/picker@2.4.2                                                                                  
npm ERR! node_modules/@react-native-picker/picker                                                                                                      
npm ERR!   @react-native-picker/picker@"2.4.2" from the root project                                                                                   
npm ERR!                                                                                                                                               
npm ERR! Fix the upstream dependency conflict, or retry                                                                                                
npm ERR! this command with --force, or --legacy-peer-deps                                                                                              
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.                                                                        
npm ERR! 
npm ERR! See /home/ecthiender/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ecthiender/.npm/_logs/2022-09-10T12_17_06_792Z-debug-0.log
```

I tried installing `2.4.4` too with `npx expo install @react-native-picker/picker@2.4.4` but I get the exact same error. Its still trying to install `2.4.2`

### What platform(s) does this occur on?

Android

### Environment

```

  expo-env-info 1.0.5 environment info:
    System:
      OS: Linux 5.19 Arch Linux
      Shell: 5.9 - /usr/bin/zsh
    Binaries:
      Node: 16.15.1 - ~/.nvm/versions/node/v16.15.1/bin/node
      Yarn: 1.22.19 - ~/.nvm/versions/node/v16.15.1/bin/yarn
      npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
    npmPackages:
      @expo/webpack-config: ^0.17.0 => 0.17.2 
      expo: ~46.0.9 => 46.0.10 
      react: 18.0.0 => 18.0.0 
      react-dom: 18.0.0 => 18.0.0 
      react-native: 0.69.5 => 0.69.5 
      react-native-web: ~0.18.7 => 0.18.9 
    npmGlobalPackages:
      expo-cli: 6.0.5
    Expo Workflow: managed

```

### Minimal reproducible example

1. Create an expo app via `npx create-expo-app my-app`
2. Install Picker via `npx expo install @react-native-picker/picker`
3. Observe the error that react dependency do not match

```
npm ERR! code ERESOLVE                                                                                                                                 
npm ERR! ERESOLVE unable to resolve dependency tree                                                                                                    
npm ERR!                                                                                                                                               
npm ERR! While resolving: my-app@1.0.0                                                                                                                 
npm ERR! Found: react@18.0.0                                                                                                                           
npm ERR! node_modules/react                                                                                                                            
npm ERR!   react@"18.0.0" from the root project                                                                                                        
npm ERR!                                                                                                                                               
npm ERR! Could not resolve dependency:                                                                                                                 
npm ERR! peer react@"16 || 17" from @react-native-picker/picker@2.4.2
npm ERR! node_modules/@react-native-picker/picker
npm ERR!   @react-native-picker/picker@"2.4.2" from the root project
npm ERR! 
````

**Comment by amandeepmittal** [comment]
Hey @ecthiender,

Seems like the package is not compatible with React Native v0.69 (for more information on this, please see the open issue in the library itself: https://github.com/react-native-picker/picker/issues/425).

The `2.4.2` is compatible with Expo SDK 46. As a workaround, you can install this package `--legacy-peer-deps` option: 

```shell
npm install @react-native-picker/picker@2.4.2 --legacy-peer-deps
```

This will ignore all peer dependencies when installing.

**Comment by hpbush** [comment]
Installing with --legacy-peer-deps works for me as well.  My project is expo and I'm in the process of migrating to EAS builds.  In order to make your builds work you need to add 
`"scripts": {
     ...
    "eas-build-pre-install": "npm config set legacy-peer-deps=true"
  },`
  to your package.json

**Comment by ecthiender** [comment]
@amandeepmittal @hpbush thanks for your info. But shouldn't the official docs for Expo be updated in this case? https://docs.expo.dev/versions/latest/sdk/picker/. It definitely looks like its an officially supported SDK component of Expo in the docs.

It seems like there are lot of things in the Expo SDK that is marketed as officially supported and part of Expo SDK, which actually are not. And if you face problems, you are often redirected to the original author's open-source project. As if its my responsibility to figure out compatibility with various components listed in Expo SDK and Expo itself! Why mark it as part of Expo SDK when its not? This is a DX nightmare. If its not officially supported, just don't mention it in the Expo docs, and I'll know I've installed a 3rd party library where I'm suppose to make sure they're compatible. Would set the expectation fair from the beginning and would save a lot of frustration and have better DX.

PS: off-topic, about Expo DX, but - 

I've had a really frustrating time to upgrade Expo from older versions and keeping my project up-to date with Expo SDKs. Often Expo will break existing apps by quickly deprecating their SDKs or having components within their SDK which doesn't work (like the one in this issue). I have a project which doesn't evolve as much. Even then I have to keep it updated every month otherwise things start breaking. And recently, one of the upgrades didn't actually work. I spent countless hours trying to upgrade from SDK 41 to 45. [It did not work](https://github.com/ecthiender/dog-e/pull/10). Finally, I had to initialize a new project with new SDK version and [manually migrate the code](https://github.com/ecthiender/dog-e/pull/13), fixing things as I do so! I don't know any other ecosystem where this is acceptable! Not at all happy with Expo :(

**Comment by ecthiender** [comment]
Btw, IIRC, setting `--legacy-peer-deps` didn't work for me. It broke a bunch of other libraries. So I had painfully recreate a half-assed version of the `Picker` in my code. 🤦🏾

**Comment by amandeepmittal** [comment]
Hey @ecthiender, I'm sorry that you are facing inconveniences with the particular library.

As I mentioned before, the library is not compatible with the current React Native (0.69+) and the [issue](https://github.com/react-native-picker/picker/issues/425) I linked does offer workarounds such as yarn which doesn't enforce peer dependencies. It's a community package and is open source.

I also reproduced a minimal example by installing the library with the version you mentioned: `npm install @react-native-picker/picker@2.4.4 --legacy-peer-deps`, and installing it works on my end. I'm curious to know how and what other libraries were affected due to not being able to install the package.

> Why mark it as part of Expo SDK when its not? This is a DX nightmare.

The reason Expo docs sometimes "recommend" a particular community package for various reasons and for folks to be able to find packages easily (especially the ones who are getting started with the mobile development ecosystem). Not every package is actively maintained by the Expo team.

> If its not officially supported, just don't mention it in the Expo docs, and I'll know I've installed a 3rd party library where I'm suppose to make sure they're compatible. Would set the expectation fair from the beginning and would save a lot of frustration and have better DX.

From Docs' perspective, we will take a look into it on how we can provide a better way of communicating for these issues.

> I've had a really frustrating time to upgrade Expo from older versions and keeping my project up-to date with Expo SDKs. 

We always release incremental upgrade options for one SDK version to another on [blog.expo.dev](https://blog.expo.dev/). I'd recommend checking it out for each SDK you are moving up if that is an inconvenience. If you think you have more to add, happy to talk about it.

> Often Expo will break existing apps by quickly deprecating their SDKs or having components within their SDK which doesn't work (like the one in this issue).

This is not a valid statement. Expo always supports a few older SDK versions for the reason that there is a breaking change in a particular third-party library that can't keep up with the changes in React Native ecosystem and your app is heavily dependent on it, you can continue to use that specific version until there is a fix for it from that third-party library. The package/library you are mentioning here is something that is related to the recent changes with React and React Native.

**Comment by ecthiender** [comment]
@amandeepmittal appreciate your response.

> The reason Expo docs sometimes "recommend" a particular community package for various reasons and for folks to be able to find packages easily (especially the ones who are getting started with the mobile development ecosystem). Not every package is actively maintained by the Expo team.

This is contradictory. You recommend packages as "officially" supported (by listing them under the Expo SDK category), yet do not take any responsibility of the compatibility. This sounds extremely off-putting to me as a developer. I find it very surprising that this stance is normalized in this community/ecosystem.

> We always release incremental upgrade options for one SDK version to another on [blog.expo.dev](https://blog.expo.dev/). I'd recommend checking it out for each SDK you are moving up if that is an inconvenience. If you think you have more to add, happy to talk about it.

I followed the docs exactly, and tried to upgrade (my PR is linked in the previous comment), but failed. Expo upgrades, maybe theoretically, are supposed to be painless and one step, but that rarely happens in practice I think. I'm working on a fairly easy but real-world project, and even then I had to struggle with so many incompatibilities even between various react native/expo packages. For example, after I upgraded SDK 41 to SDK 42, I was informed by Expo the following react-native packages were incompatible ` @react-native-community/masked-view, @react-navigation/bottom-tabs, @react-navigation/native, @react-navigation/stack`. So I had to manually go through each of them and upgrade them individually and fixing errors.  And this obviously happens with a large no. of third-party packages. 

> This is not a valid statement. Expo always supports a few older SDK versions for the reason that there is a breaking change in a particular third-party library that can't keep up with the changes in React Native ecosystem and your app is heavily dependent on it, you can continue to use that specific version until there is a fix for it from that third-party library. The package/library you are mentioning here is something that is related to the recent changes with React and React Native.

From what I've noticed, usually a SDK version is deprecated in 8-12 months. So that means I am forced to go through this process atleast once a year. Now, I understand most commercial projects move very fast so this not a problem for most projects. But I work on a project which doesn't change very frequently. But supporting the app is important. So I have to keep upgrading my app (and spend unnecessary dev effort) every 2-3 months just to keep it updated. This is necessary because upgrading once a year and jumping few SDK versions causes more issues. So I am kind of forced to update every SDK release. 

TL;DR - I have worked in various JS ecosystems, including React. Worked in Python, Haskell and Golang. While none of these ecosystems are perfect, I have never faced this kind of situation, where to keep my app up-to-date I have to spend more effort than regular dev effort and the process is extremely brittle and error-prone, and ultimately the whole thing is very frustrating. This is just my experience and opinion. Maybe most devs have a great time working with Expo, I don't know. But I, certainly, did not enjoy working with Expo. Just my 2 cents.

**Comment by amandeepmittal** [comment]
@ecthiender, thanks for the verbose and in-detail feedback! Really appreciate you took time to write and share the pain points.

I know some people at Expo maintain apps like yours that are important but are not being developed every day/week. It definitely makes upgrading hard when packages just don’t work. I understand your frustration, especially your expectation of quality and how much effort you expected to put forth was broken. One of our goals is to make SDK releases break fewer things, but we aren’t there yet. We are also working on making our docs better.

Moving forward, I’ll talk to the team about auditing these packages/what we include under the Expo SDK in our docs. I know this hasn’t been a fun process, and we appreciate that you told us you were having an issue 🙏
